akka-docs/src/main/paradox/scala/stream/stream-substream.md

```diff
@@ -0,0 +1,125 @@
+during the runtime of the flow graph processing.
+
+Stages that creates substreams are listed on @ref:[Nesting and flattening stages](stages-overview.md#nesting-and-flattening-stages)
+
+## Nesting stages
+
+### groupBy
+
+A typical operation that generates substreams is `groupBy`.
+
+Scala
+:   @@snip [SubstreamDocSpec.scala]($code$/scala/docs/stream/SubstreamDocSpec.scala) { #groupBy1 }
+
+Java
+:   @@snip [SubstreamDocTest.java]($code$/java/jdocs/stream/SubstreamDocTest.java) { #groupBy1 }
+
+![stream-substream-groupBy1.png](../../images/stream-substream-groupBy1.png)
+
+This operation splits the incoming stream into separate output
+streams, one for each element key. The key is computed for each element
+using the given function, which is `f` in the above diagram. When a new key is encountered for the first time
+a new substream is opened and subsequently fed with all elements belonging to that key.
+
+If you add a `Sink` or `Flow` right after the `groupBy` stage,
+all transformations are applied to all encountered substreams in the same fashion.
+So, if you add the following `Sink`, that is added to each of the substreams as in the below diagram.
+
+Scala
+:   @@snip [SubstreamDocSpec.scala]($code$/scala/docs/stream/SubstreamDocSpec.scala) { #groupBy2 }
+
+Java
+:   @@snip [SubstreamDocTest.java]($code$/java/jdocs/stream/SubstreamDocTest.java) { #groupBy2 }
+
+![stream-substream-groupBy2.png](../../images/stream-substream-groupBy2.png)
+
+Also substreams, more precisely, `SubFlow` and `SubSource` have methods that allow you to
+merge or concat substreams into the master stream again.
+
+The `mergeSubstreams` method merges an unbounded number of substreams back to the master stream.
+
+Scala
+:   @@snip [SubstreamDocSpec.scala]($code$/scala/docs/stream/SubstreamDocSpec.scala) { #groupBy3 }
+
+Java
+:   @@snip [SubstreamDocTest.java]($code$/java/jdocs/stream/SubstreamDocTest.java) { #groupBy3 }
+
+![stream-substream-groupBy3.png](../../images/stream-substream-groupBy3.png)
+
+You can limit the number of active substreams running and being merged at a time,
+with either the `mergeSubstreamsWithParallelism` or `concatSubstreams` method.
+
+Scala
+:   @@snip [SubstreamDocSpec.scala]($code$/scala/docs/stream/SubstreamDocSpec.scala) { #groupBy4 }
+
+Java
+:   @@snip [SubstreamDocTest.java]($code$/java/jdocs/stream/SubstreamDocTest.java) { #groupBy4 }
+
+However, since the number of active substreams is capped,
+be careful so that these methods do not cause deadlocks with back pressure like in the below diagram.
+
+![stream-substream-groupBy4.png](../../images/stream-substream-groupBy4.png)
+
+### splitWhen and splitAfter
+
+`splitWhen` and `splitAfter` are two other operations which generate substreams.
+
+The difference from `groupBy` is that, if the predicate for `splitWhen` and `splitAfter` returns true,
+a new substream is generated, and the succeeding elements after split will flow into the new substream.
```

Comment:
I wonder if we could come up with an example use case when this is useful as well?

"This is useful for example for ..."

Can't think of anything obvious from the top of my head though.

Reply:
Sorry ... can't think of a good example either :|

Reply:
@akka/akka-team does anyone else have a good idea for this section?

Reply:
For split rather than group? Any time you know once you've scanned over something you don't care about anything behind it e.g. counting the number of words per line in a stream:

```
  val lineCount = words.splitAfter(_ == "\n")
    .filter(_ != "\n")
    .map(_ => 1)
    .reduce(_ + _)
```